### ISSUE TITLE
DFlash: implementation gaps vs. paper likely behind ~2% acceptance

### ISSUE THREAD

**Issue by dpalmqvist** [motsognirr/olmlx#317]
CLAUDE.md notes that locally-trained DFlash drafts hit ~2% acceptance, far below the paper's reported numbers ([arxiv:2602.06036](https://arxiv.org/abs/2602.06036), 6× speedup / >2.5× over EAGLE-3). I went through `olmlx/engine/dflash/` against the upstream MLX reference (`z-lab/dflash`, `dflash/model_mlx.py`) and the paper writeup. The decoder/architecture is paper-faithful enough to load z-lab pre-trained drafts, but the **trainer** has divergences that plausibly explain the gap.

## What's faithful

- Cross-attention with proposal/context streams; only ctx K/V written to cache (`olmlx/engine/dflash/draft_model.py:158-227` matches `dflash/model_mlx.py:82-116`)
- `fc + hidden_norm` projection of concatenated target hidden states
- `bind()` borrowing `embed_tokens`/`lm_head`
- Bidirectional within-block attention (`attention_causal=False`)
- Greedy verify + bonus token; GDN rollback for hybrid linear-attention targets
- Universal target hook via `_LayerHook` / `_patch_model`
- **Inference RoPE alignment is correct**: `_hidden` covers ctx positions only, pending sits at `cache.offset + S`

## Gap 1 — train/test slice mismatch (off-by-one), most likely root cause

`olmlx/engine/dflash/prepare.py:1003`:

```python
target_hidden = target_hidden_full[:, : p + 1, :]
```

This includes position `p` (the pending token's own position) in ctx. At inference (`olmlx/engine/dflash/decoder.py`), ctx never contains the pending's position — `_hidden` is sliced to `num_accepted` from the previous verify and the new pending sits at the position immediately after.

|            | ctx covers | pending RoPE pos | pending content pos |
|------------|------------|------------------|---------------------|
| Training   | 0..p       | p+1              | p                   |
| Inference  | 0..L−1     | L                | L                   |

Two consequences:
1. During training the draft sees the **target's own hidden for the pending token** in ctx — extra information not available at inference. The draft can learn to copy from that ctx hidden; that path is dead at inference.
2. Proposal tokens are trained at RoPE positions `p+1..p+block_size+1` while inference uses `L..L+block_size`. RoPE is relative-position, so the *attention math* survives, but the *ctx-content distribution* shifts.

**Fix**: slice `[:, :p, :]` instead of `[:, :p+1, :]`. One-line change worth trying first.

## Gap 2 — no per-position loss weighting

The paper trains with `w_k = exp(-(k-1)/γ)` to emphasize early positions because acceptance length compounds — position 1 wrong wastes positions 2..16 (the cascade effect). `olmlx/engine/dflash/prepare.py:397` uses uniform mean:

```python
ce = mx.mean(nll)
```

Without this weighting, the optimizer spends the same gradient budget on later positions even though they only matter conditional on every earlier position being accepted. Depresses acceptance length even when token-1 accuracy is comparable.

**Fix**: multiply per-position NLL by `exp(-(k-1)/γ)` before reduction; `γ` becomes a hyperparameter (paper doesn't pin a single value publicly — start with γ=block_size/2 and sweep).

## Gap 3 — defaults don't match paper's reported configuration

| Setting             | olmlx default | Paper |
|---------------------|---------------|-------|
| Block size          | 4             | 16    |
| Draft layers        | 4             | 5     |
| Target hidden layers| 4             | 5     |
| Steps               | 2000          | many more |

In `olmlx/engine/dflash/prepare.py:64-66` and `olmlx/engine/dflash/decoder.py:212` (`DFlashDecoder.__init__(block_size=4)` is the inference-side default if a trained draft's `block_size` isn't propagated through).

## Gap 4 — layer-selection algorithm differs

Upstream `build_target_layer_ids(N, K)` picks indices in `[1, N−3]` (early-biased, e.g. `[1,8,15,22,29]` for N=32, K=5). `_evenly_spaced` in `olmlx/engine/dflash/prepare.py:106-142` picks center-biased indices (e.g. `[5,10,16,21,26]`). Self-consistent for olmlx-trained drafts (the chosen ids get persisted to `config.json`), but probably suboptimal — the paper's recipe weights early/middle layers specifically.

## Gap 5 — one window per sequence per step

Documented in the prepare.py module docstring as a known under-utilization vs. the paper's "fixed number of masked blocks per sequence, anchors resampled each epoch."

## Gap 6 — sliding-attention layers in non-causal mode get no spatial mask

`olmlx/engine/dflash/draft_model.py:217-221`:
```python
mask = "causal" if self.causal else None
if self.is_sliding and self.causal and ctx_len + L > (self.sliding_window or 0):
    mask = create_causal_mask(L, offset=ctx_len, window_size=self.sliding_window)
```

Upstream MLX always applies a sliding-causal mask on sliding layers regardless of causal mode (`dflash/model_mlx.py:110-114`). Doesn't bite Qwen3/3.5 (all `full_attention`); would bite Gemma 4-style draft architectures.

## Recommended order of attack

1. **Gap 1** (one-line slice fix) — verify with a small training run on a Qwen3 target and `olmlx bench`.
2. **Gap 3** — bump `DEFAULT_BLOCK_SIZE` to 16 and `DEFAULT_NUM_HIDDEN_LAYERS`/target layers to 5 to match paper.
3. **Gap 2** (per-position weighting) — measure delta after gaps 1 and 3.
4. **Gap 4** — port upstream's `build_target_layer_ids` for parity.
5. Re-bench against the empirical-pick claim in CLAUDE.md ("classic gets 1.33–1.92× on Qwen3.5"); if DFlash now competes, update the recommendation.

Filed to track investigation; not committing to ship any of these without bench evidence.

**Comment by dpalmqvist** [comment]
Gaps 1-4 were already implemented in the codebase before this investigation.

- **Gap 1** (slice mismatch): Already fixed — `target_hidden = target_hidden_full[:, :p, :]` at prepare.py:1093
- **Gap 2** (per-position weighting): Already implemented as `position_decay_gamma` parameter in `_draft_loss`
- **Gap 3** (defaults): Already bumped to `DEFAULT_NUM_HIDDEN_LAYERS=5`, `DEFAULT_NUM_TARGET_LAYERS=5`, `DEFAULT_BLOCK_SIZE=16`
- **Gap 4** (layer selection): Already ported — `_evenly_spaced` now follows upstream's `[1, N-3]` range with early bias
- **Gap 5** (one window per sequence): Documented as a known structural limitation in the module docstring; multi-window training would be a separate feature
- **Gap 6** (sliding attention mask): Fixed — removed the `self.causal` guard from the sliding condition in `DFlashAttention.__call__`, matching upstream z-lab/dflash behavior

All 3086 tests pass.
